### ISSUE TITLE
MakeImmutable doesn't always move objects into the public region

### ISSUE THREAD

**Issue by ssiccha** [gap-system/gap#3117]
If I create e.g. a compressed matrix and call `MakeImmutable` on it, it should be moved to the public region but isn't.
On current master:
```
gap> echelon := RandomMat(20, 20, GF(5));;
gap> ConvertToMatrixRepNC(echelon);;
gap> echelon;
< mutable compressed matrix 20x20 over GF(5) >
gap> MakeImmutable(echelon);
< immutable compressed matrix 20x20 over GF(5) >
gap> IsPublic(echelon);
false
```

**Comment by ssiccha** [comment]
The hpc-gap manual states:
> When MakeImmutable (Reference: MakeImmutable) is used, its argument is automatically moved to the public region.

@fingolfin told me that per default in HPC-GAP component objects are atomic, but positional objects are not. And only atomic component or positional objects can be moved into the public region.

**Comment by rbehrends** [comment]
The manual is incomplete here, it appears. `MakeImmutable()` works this way for *most* objects, but for component and positional objects, making them immutable only makes them *mathematically* immutable; as they may still contain writable data, the kernel cannot safely move them to the public region. The workaround is to use `MakeReadOnlyObj()` instead.

**Comment by ssiccha** [comment]
@rbehrends Could there _in principle_ be a clean way to turn a non atomic positional or component object into an atomic one? Or would the safer and cleaner solution be to make an atomic copy?

**Comment by rbehrends** [comment]
The problem with atomic objects is that you have to ensure yourself that no race conditions occur. I.e. if two threads modify the object at the same time, the only thing that the kernel guarantees is that it won't lead to a crash. But if you have concurrent updates, it's basically random, which thread "wins".

That said, you can use the `IsAtomicPositionalObjectRep` to turn atomicity on for positional objects and `IsNonAtomicComponentObjectRep` to turn it off for component objects if you know what you are doing.

**Comment by stevelinton** [comment]
`MakeReadOnly` is probably the answer to what you need here. In actual fact, I know of no methods that write to the data structure underlying an immutable compressed matrix, but there are other similar data structures that do get written to. For complex objects like groups, HPCGAP will use the atomic component object representation, since (at least most of the time) attribute values are mathematically defined, so any stored value will be OK.   It might be possible to use PostMakeImmutable methods to have some positional objects become readonly when they are made immutable.

**Comment by ssiccha** [comment]
@stevelinton You're right, I've ended up using `MakeReadOnly` and in my case this is what I actually wanted.

I had a short chat with @fingolfin and @markuspf yesterday and it looks like there nevertheless is a case that needs to be adressed:
If somebody creates a non-atomic component or positional object there is no (documented) way that they can move their object to the public region. If they want to keep their object in `IsAttributeStoringRep` they also can't move their object into the read-only region.

If there is a way to turn a non-atomic component or positional object into an atomic one, then I would suggest that `MakeImmutable` should do the following:
If its argument is a non atomic component or positional object it is made immutable and then converted into an atomic one.

I would find it more consistent if `MakeImmutable` really moved _all_ objects into the public region. What do you think?

Independently of that, the manual could be updated to explicitly say that:
- access to an object in the read-only region is more efficient and safer than access to an atomic object.
- since the attributes of objects in the public region are not necessarily immutable the access to these _attributes_ can be subject to race conditions.

**Comment by stevelinton** [comment]
There is no way, in general, to be sure that it makes any kind of sense for a non-readonly object to be in 
the public region.  It is quite legitimate for a mathematically immutable object, as part of  some method called on it, to be converted to a completely different representation (still with the same mathematical meaning), giving rise to arbitrarily unpleasant race conditions, much nastier than the ones that arise from Attribute Storing. 

Objects in `IsAttributeStoringRep` are component objects, and will normally be immutable and atomic anyway (and yes, race conditions involving multiple attributes with related values are a possible problem -- the solution here is to get rid of such attributes.

For non attribute storing objects (typically positional objects, representing elements rather than domains) it will often not make sense to convert them to atomic -- updating one field of the positional object is not especially likely to be a sensible atomic operation. If they can't be made read-only then either (a) if you know atomic does make sense, and you create the objects then you can make them atomic at that stage or (b) you need to put them in a shared region and manage access explicitly.  

In short, it simply doesn't make sense for some objects to be in the public region. 

Clearly documentation needs work.

**Comment by fingolfin** [comment]
BTW, folks, please note that read-only objects are *not* in the public region either (they are in the read-only region instead). So, `IsPublic` returns `false` for them. That's somewhat confusing (and I wonder if all code gets that right...)
